arangod/Transaction/Methods.cpp

```diff
@@ -1148,151 +1273,169 @@ Future<OperationResult> transaction::Methods::insertLocal(
+          buildDocumentIdentity(*collection, resultBuilder, cid,
                                 key.stringView(), oldRevisionId,
                                 RevisionId::none(), nullptr, nullptr);
         }
-        // we have not written anything, so exclude this document from
-        // replication!
-        excludeFromReplication = true;
-
-#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-        numExclusions++;
-#endif
-
         return res;
       }
 
       if (options.overwriteMode == OperationOptions::OverwriteMode::Update) {
         // in case of unique constraint violation: (partially) update existing
-        // document
-        res =
-            collection->update(this, value, docResult, options, prevDocResult);
-        if (res.ok() && prevDocResult.revisionId() == docResult.revisionId()) {
-          excludeFromReplication = true;
+        // document.
+        previousDocumentBuilder->clear();
+        res = collection->getPhysical()->lookupDocument(
+            *this, oldDocumentId, *previousDocumentBuilder,
+            /*readCache*/ true, /*fillCache*/ false, ReadOwnWrites::yes);
 
-#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-          numExclusions++;
-#endif
+        if (res.ok()) {
+          TRI_ASSERT(previousDocumentBuilder->slice().isObject());
+
+          res = modifyLocalHelper(
+              *collection, value, oldDocumentId, oldRevisionId,
+              previousDocumentBuilder->slice(), newRevisionId,
+              *newDocumentBuilder, options, batchOptions, /*isUpdate*/ true);
+        }
+
+        TRI_ASSERT(res.fail() || newDocumentBuilder->slice().isObject());
+
+        if (res.ok() && oldRevisionId == newRevisionId) {
+          // did not actually update - intentionally do not fill replicationData
+          excludeFromReplication |= true;
         }
       } else if (options.overwriteMode ==
                  OperationOptions::OverwriteMode::Replace) {
-        // in case of unique constraint violation: replace existing document
-        // this is also the default behavior
-        res =
-            collection->replace(this, value, docResult, options, prevDocResult);
+        // in case of unique constraint violation: replace existing document.
+        // this is also the default behavior for "overwrite: true".
+        previousDocumentBuilder->clear();
+        res = collection->getPhysical()->lookupDocument(
+            *this, oldDocumentId, *previousDocumentBuilder,
+            /*readCache*/ true, /*fillCache*/ false, ReadOwnWrites::yes);
+
+        if (res.ok()) {
+          TRI_ASSERT(previousDocumentBuilder->slice().isObject());
+
+          res = modifyLocalHelper(
+              *collection, value, oldDocumentId, oldRevisionId,
+              previousDocumentBuilder->slice(), newRevisionId,
+              *newDocumentBuilder, options, batchOptions, /*isUpdate*/ false);
+        }
+
+        TRI_ASSERT(res.fail() || newDocumentBuilder->slice().isObject());
```

Comment:
The two code paths look almost identical except for the `/*isUpdate* true` vs `/*isUpdate* false` and the handling of `excludeFromReplication`. Can we streamline this and remove duplicate code?